### PR TITLE
magento/magento2#39863: Magento Catalog Product list widget returns incorrect count

### LINKED ISSUE #39863: Magento Catalog Product list widget returns incorrect count
### Preconditions and environment

- Magento version - 2.4.7-p4

### Steps to reproduce

1. insert a catalog product widget from page builder  (Insert Row, drag Html Code and insert widget)
2. set no  product per page = 5
3. set no of product to display = 9

### Expected result

a widget should show total 9 product with 5 products on PAGE 1 and 4 products on page 2

### Actual result

instead it display total 10 product 5 on PAGE 1 and 5 ON PAGE 2

### Additional information

_No response_

### Release note

_No response_

### Triage and priority

- [x] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [x] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by MaximLogic**
### Description (*)

According to issue #39863, the catalog product widget with pagination enabled was displaying more products than configured when the total products count is not a multiple of products per page. For example, when configuring the widget with `products_per_page = 5` and `products_count = 9`, the widget displayed 10 products total (5 on page 1, 5 on page 2) instead of the expected 9 products (5 on page 1, 4 on page 2). Additionally, the pager showed incorrect text "Items 6 to 10 of 9 total" on the second page.
The problem occurring, because in `ProductsList::getBaseCollection()`, the collection limit was set using `setPageSize()` and `setCurPage()` methods without considering the `products_count` (total limit) constraint. This caused the SQL query to fetch full page size worth of products regardless of the total limit.

**Solution:**
I replaced automatic pagination with manual SQL LIMIT calculation in `getBaseCollection()`, and added logic to calculate the correct limit: `min(pageSize, max(0, totalLimit - offset))`. This ensures that on the last page, only the remaining products within the total limit are fetched. Also, updated unit tests to reflect the new behavior and added test cases for edge scenarios

### Fixed Issues (if relevant)

1. Fixes magento/magento2#39863

### Manual testing scenarios (*)

1. Login to Magento Admin
2. Navigate to Content → Pages
3. Select any existing page or create a new page
4. Click "Edit with Page Builder" (if you use Text Editor, skip this and next step)
5. Drag and drop a Row element, then drag HTML Code inside it
6. Click "Insert Widget" button
7. Select "Catalog Products List" widget type
8. Configure the widget:
   - Set "Number of Products per Page" = 5
   - Set "Number of Products to Display" = 9
   - Add any condition to filter products (e.g., category)
   - Save the widget configuration
9. Save the page and view it on the frontend
10. Verify page 1 displays exactly 5 products
11. Click the pagination to go to page 2
12. **Expected result:** Page 2 displays exactly 4 products (not 5)
13. **Expected result:** Pager shows "Items 6 to 9 of 9 total"
14. **Expected result:** Total of 9 products displayed across both pages

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by MaximLogic** [comment]
@magento run all tests

**Comment by engcom-Hotel** [comment]
@magento run all tests

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php [hunk 1]
```diff
@@ -1,6 +1,6 @@
 <?php
 /**
- * Copyright 2014 Adobe
+ * Copyright 2026 Adobe
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogWidget/Block/Product/ProductsList.php]: Keep it `2014` only

File: app/code/Magento/CatalogWidget/Test/Unit/Block/Product/ProductsListTest.php [hunk 2]
```diff
@@ -1,6 +1,6 @@
 <?php
 /**
- * Copyright 2015 Adobe
+ * Copyright 2026 Adobe
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogWidget/Test/Unit/Block/Product/ProductsListTest.php]: Keep it `2015`

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php [hunk 3]
```diff
@@ -397,6 +395,21 @@ public function getBaseCollection(): Collection
          */
         $collection->distinct(true);
 
+        $currentPage = (int)$this->getRequest()->getParam($this->getData('page_var_name'), 1);
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogWidget/Block/Product/ProductsList.php]: Missing null safety. If `getParam()` returns `null`, casting to `int` returns `0`, but the default is expected to be `1`.

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php [hunk 4]
```diff
@@ -383,9 +383,7 @@ public function getBaseCollection(): Collection
         $collection = $this->_addProductAttributesAndPrices($collection)
             ->addStoreFilter()
             ->addAttributeToFilter(Product::STATUS, ProductStatus::STATUS_ENABLED)
-            ->addAttributeToSort('entity_id', 'desc')
-            ->setPageSize($this->getPageSize())
-            ->setCurPage($this->getRequest()->getParam($this->getData('page_var_name'), 1));
+            ->addAttributeToSort('entity_id', 'desc');
```
Comment (engcom-Hotel) [inline app/code/Magento/CatalogWidget/Block/Product/ProductsList.php:386]: The switch from `setPageSize()`/`setCurPage()` to `getSelect()->limit()` bypasses Collection's internal pagination state. This might breaks:

- `$collection->getCurPage()` → Always returns 1
- `$collection->getPageSize()` → Always returns 0  
- `$collection->getLastPageNumber()` → Always returns 1
- Pager block relies on these methods

**Suggestion:** Use Collection methods to maintain backward compatibility:
$collection->setPageSize($effectivePageSize);
$collection->setCurPage($currentPage);

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php (hunk 4 above)
Comment (MaximLogic) [reply app/code/Magento/CatalogWidget/Block/Product/ProductsList.php:386]: Well, in that case, the issue cannot be resolved, because it is collection bug, not slider's

Comment (engcom-Hotel) [reply app/code/Magento/CatalogWidget/Block/Product/ProductsList.php:386]: It seems you’re correct that the issue can’t be fixed with only `setPageSize()` / `setCurPage()`: the Framework collection has no concept of a “total cap” (e.g. “at most 9 products”), so we do need to apply the correct limit/offset via the select (e.g. getSelect()->limit($limit, $offset)) to fix the wrong count. I’m fine with that approach.

But please fix the other review comments.

Comment (engcom-Hotel) [reply app/code/Magento/CatalogWidget/Block/Product/ProductsList.php:386]: Please refer to the failed tests as well

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php (hunk 3 above)
Comment (MaximLogic) [reply app/code/Magento/CatalogWidget/Block/Product/ProductsList.php]: Done

File: app/code/Magento/CatalogWidget/Test/Unit/Block/Product/ProductsListTest.php (hunk 2 above)
Comment (MaximLogic) [reply app/code/Magento/CatalogWidget/Test/Unit/Block/Product/ProductsListTest.php]: Done

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php (hunk 1 above)
Comment (MaximLogic) [reply app/code/Magento/CatalogWidget/Block/Product/ProductsList.php]: Done

File: app/code/Magento/CatalogWidget/Block/Product/ProductsList.php (hunk 4 above)
Comment (MaximLogic) [reply app/code/Magento/CatalogWidget/Block/Product/ProductsList.php:386]: Fixed all other review comments, regarding failed tests, it looks like those aren't related to the PR

**Comment by MaximLogic** [comment]
@magento run all tests

**Review by engcom-Hotel — CHANGES_REQUESTED** [review]
Thank you @MaximLogic for making the changes.

Please have a look into the static test failures as well.

**Review by engcom-Hotel — APPROVED** [review]
Thank you @MaximLogic for making the changes. Approving this PR for further processing.

**Comment by engcom-Dash** [comment]
@magento run all tests

**Comment by engcom-Dash** [comment]
@magento run Integration Tests, Unit Tests
